### PR TITLE
fix(combobox): PlaceholderText not displaying

### PR DESCRIPTION
**Pull request by Xiaoy312**
GitHub Issue (If applicable): #3358

## PR Type
What kind of change does this PR introduce?
- Bugfix

## What is the new behavior?
ComboBox.PlaceholderText is now displaying correctly.

## PR Checklist
Please check if your PR fulfills the following requirements:
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information
Internal Issue (If applicable):
https://dev.azure.com/nventive/Uno%20Gallery/_queries/edit/191190

### CODE REVIEW

**Comment by agneszitte-nventive** [comment]
You can precise "Fix #3358" in the PR description to close the issue automatically when it's merged.
Also do you think that it should be a good idea to add some additional Uni tests / UI tests for the placeholder ;) ?

File: src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs [hunk 1]
```diff
@@ -250,36 +243,59 @@ private void UpdateContentPresenter()
...
+						}
+					}
+					else
 					{
-						// Keep track of the former parent, so we can put the item back when the dropdown is shown
-						_selectionParentInDropdown = (itemView.GetVisualTreeParent() as IWeakReferenceProvider)?.WeakReference;
+						_selectionParentInDropdown = null;
+					}
+
+					_contentPresenter.Content = item ?? _placeholderTextBlock;
```
[the remark this answers is not in the document]
Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:273]: uwp doesnt seem to use ItemTemplate for PlaceholderText:

image
```xml
		<ComboBox x:Name="TestBox3" PlaceholderText="PlaceholderText" ItemsSource="asda sdsd">
			<ComboBox.ItemTemplate>
				<DataTemplate>
					<StackPanel Orientation="Horizontal">
						<Rectangle Height="12" Width="12" Fill="Pink" />
						<TextBlock Text="{Binding}" />
					</StackPanel>
				</DataTemplate>
			</ComboBox.ItemTemplate>
		</ComboBox>
```

File: src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs [hunk 2]
```diff
@@ -250,36 +243,59 @@ private void UpdateContentPresenter()
...
-
-				if (itemView != null)
+				if (SelectedItem != null)
 				{
+					var item = GetSelectionContent();
+					if (item is _View itemView)
+					{
 #if __ANDROID__
-					var comboBoxItem = itemView.FindFirstParentOfView<ComboBoxItem>();
+						var comboBoxItem = itemView.FindFirstParentOfView<ComboBoxItem>();
```
[the remark this answers is not in the document]
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:252]: `SelectedItem` may *not* be a `ComboBoxItem` though, in fact it usually won't be. Also it's not an ironclad guarantee that `ContentControl.Content` will be a child of `ContentControl`.

File: src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs [hunk 3]
```diff
@@ -250,36 +243,59 @@ private void UpdateContentPresenter()
...
+						if (comboBoxItem != null)
+						{
+							// Keep track of the former parent, so we can put the item back when the dropdown is shown
+							_selectionParentInDropdown = (itemView.GetVisualTreeParent() as IWeakReferenceProvider)?.WeakReference;
+						}
+
+						if (itemView.GetVisualTreeParent() != _contentPresenter)
+						{
+							// Item may have been put back in list, reattach it to _contentPresenter
+							_contentPresenter.AddChild(itemView);
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:265]: Why did you change the timing of this line?

File: src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:252]: I tried to avoid walking up the tree when the item is ComboBoxItem:
```cs
        private ComboBoxItem GetSelectionContainer()
        {
            if (SelectedItem is ComboBoxItem item)
            {
                return item;
            }
            else if (SelectedItem is _View view)
            {
#if __ANDROID__
                return view.FindFirstParentOfView<ComboBoxItem>();
#else
                return view.FindFirstParent<ComboBoxItem>();
#endif
            }
            else
            {
                return null;
            }
        }
        private object GetSelectionContent()
        {
            return SelectedItem is ComboBoxItem cbi ? cbi.Content : SelectedItem;
        }
//...
var item = GetSelectionContent();
if (item is _View itemView)
{
   var comboBoxItem = itemView.FindFirstParentOfView<ComboBoxItem>(); // old impl
   var comboBoxItem2 = GetSelectionContainer(); // new impl attempt
```

However that lead to some issue when the item got moved under ContentPresenter in the visual tree. 
Particularly, in the "ComboBox_ComboBoxItem_Sel..." sample, when "item 1" is selected and you open the flyout again, the item 1 is not restored. (Item 2-4 didn't have this problem.) What I noticed is that `UpdateContentPresenter` was called twice, and on the second call comboBoxItem is null, while comboBoxItem2 has value. Maybe this corrupts the `_selectionParentInDropdown`, thus prevent the item in the dropdown being restored.

I am leaving the code as it is now.

File: src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.cs [hunk 4]
```diff
@@ -697,7 +703,9 @@ private void SynchronizeContentTemplatedParent()
 				{
 					var templatedParent = _contentTemplateRoot is ImplicitTextBlock
 						? this // ImplicitTextBlock is a special case that requires its TemplatedParent to be the ContentPresenter
-						: (this.TemplatedParent as IFrameworkElement)?.TemplatedParent;
+						: !SynchronizeContentWithOuterTemplatedParent && _dataTemplateUsedLastUpdate == null
```
[the remark this answers is not in the document]
Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.cs]: Yeah, that part wasn't clear at first glance. I've added a comment to clarify that.

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ComboBoxTests/UnoSamples_Tests.ComboBoxTests.cs [hunk 5]
```diff
@@ -214,5 +214,67 @@ public void ComboBoxTests_ToggleDisabled()
 
 			TakeScreenshot("ComboBox Disabled");
 		}
+
+		[Test]
+		[AutoRetry]
+		[ActivePlatforms(Platform.Android)]
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ComboBoxTests/UnoSamples_Tests.ComboBoxTests.cs:220]: Why only android ?
